piq/pieapp.py

```diff
@@ -0,0 +1,192 @@
+            x11: Outputs of the last convolutional layer used as weights
+        """
+        _validate_input(input_tensors=x, allow_5d=False, allow_negative=False)
+        x = _adjust_dimensions(input_tensors=x)
+        assert x.shape[2] == x.shape[3] == self.FEATURES, \
+            f"Expected square input with shape {self.FEATURES, self.FEATURES}, got {x.shape}"
+
+        # conv1 -> relu -> conv2 -> relu -> pool -> conv3 -> relu
+        x3 = F.relu(self.conv3(self.pool(F.relu(self.conv2(F.relu(self.conv1(x)))))))
+        # conv4 -> relu -> pool -> conv5 -> relu
+        x5 = F.relu(self.conv5(self.pool(F.relu(self.conv4(x3)))))
+        # conv6 -> relu -> pool -> conv7 -> relu
+        x7 = F.relu(self.conv7(self.pool(F.relu(self.conv6(x5)))))
+        # conv8 -> relu -> pool -> conv9 -> relu
+        x9 = F.relu(self.conv9(self.pool(F.relu(self.conv8(x7)))))
+        # conv10 -> relu -> pool1-> conv11 -> relU
+        x11 = self.flatten(F.relu(self.conv11(self.pool(F.relu(self.conv10(x9))))))
+        # flatten and concatenate
+        features = torch.cat((self.flatten(x3), self.flatten(x5), self.flatten(x7), self.flatten(x9), x11), dim=1)
+        return features, x11
+
+    def compute_difference(self, features_diff, weights_diff):
+        r"""
+        Args:
+            features_diff: Tensor of shape (N, C_1)
+            weights_diff: Tensor of shape (N, C_2)
+        Returns:
+            distances
+            weights
+        """
+        # Get scores: fc1_score -> relu -> fc2_score
+        # 0.01 is the sigmoid coefficient
+        distances = self.ref_score_subtract(0.01 * self.fc2_score(F.relu(self.fc1_score(features_diff))))
+
+        weights = self.fc2_weight(F.relu(self.fc1_weight(weights_diff))) + self.EPS
+        return distances, weights
+
+
+class PieAPP(_Loss):
+    r"""
+    Implementation of Perceptual Image-Error Assessment through Pairwise Preference.
+    
+    Expects input to be in range [0, 1] with no normalization and RGB channel order.
+    Input images are croped into smaller patches. Score for each individual image is mean of it's patch scores.
+
+    Args:
+        reduction: Reduction over samples in batch: "mean"|"sum"|"none".
+        data_range: Value range of input images (usually 1.0 or 255). Default: 1.0
+        stride: Step between cropped patches. Smaller values lead to better quality,
+            but cause higher memory consumption. Default: 27 (`sparse` sampling in original implementation)
+        enable_grad: Flag to compute gradients. Usefull when PieAPP used as a loss. Default: False.
+    
+    References:
+        .. [1] Ekta Prashnani, Hong Cai, Yasamin Mostofi, Pradeep Sen
+            (2018). PieAPP: Perceptual Image-Error Assessment through Pairwise Preference
+            https://arxiv.org/abs/1806.02067
+        .. [2] https://github.com/prashnani/PerceptualImageError
+
+    """
+    # TODO(zakajd) 10/12/2020: Load weights to release and change this link
```

Comment:
@snk4tr this PR is ready for merge if you don't have additional comments.

I'm currently in the place with very slow internet, so it would be great if you also upload model weights to last release and change the link in code.

Reply:
Done